### PR TITLE
Purge deploys using latest block timestamp

### LINKED ISSUE #2155: Purge expired deploys from block proposer using latest block timestamp
The `BlockProposerDeploySets::prune` takes a current timestamp as means of deciding which deploys have expired.  Rather than using `Timestamp::now()` to provide this value, we should use the `timestamp` from the latest finalized block.

### PR DESCRIPTION
**Pull request by darthsiroftardis**
CHANGELOG:

- Changed the pruning functionality in the `block_proposer` to use the timestamp of the latest finalized block as seen by the block proposer

Closes #2155

### CODE REVIEW

File: node/src/components/block_proposer.rs [hunk 1]
```diff
@@ -269,7 +274,7 @@ impl BlockProposerReady {
                     .event(|_| Event::Prune);
 
                 // Announce pruned hashes
-                let pruned_hashes = self.prune(Timestamp::now());
+                let pruned_hashes = self.prune(self.last_finalized_timestamp);
```
Comment (afck) [inline node/src/components/block_proposer.rs]: Should we get rid of the `PRUNE_INTERVAL` and `Event::Prune` now, and just prune whenever `last_finalized_timestamp` changes?

**Review by afck — COMMENTED** [review]
We should add checks to both call sites to make sure we only `propose_block_payload` if `request.context.timestamp() >= self.last_finalized_timestamp`, and otherwise `warn!`.
It's probably only relevant if our clock is wildly out of sync, but it's better to be safe and make sure we never propose an invalid block.

**Review by rafal-ch — APPROVED** [review]
LGTM

File: node/src/components/block_proposer.rs (hunk 1 above)
Comment (darthsiroftardis) [reply node/src/components/block_proposer.rs]: Removed `Event::Prune`

**Review by goral09 — APPROVED** [review]

File: node/src/components/block_proposer.rs [hunk 2]
```diff
@@ -261,38 +249,33 @@ impl BlockProposerReady {
                             request.random_bit,
                         ))
                         .ignore()
+                } else {
+                    Effects::new()
```
Comment (afck) [inline node/src/components/block_proposer.rs:253]: We should log a warning here. It probably means that a block request spent so much time in the reactor queue (or in the block proposer queue?) that in the meantime we finalized a block with a timestamp newer than this.

File: node/src/components/block_proposer.rs [hunk 3]
```diff
@@ -261,38 +249,33 @@ impl BlockProposerReady {
...
                 Effects::new()
             }
             Event::FinalizedBlock(block) => {
+                let mut effects = Effects::new();
                 // We have received a finalized block, set the timestamp to
                 // allow for pruning.
                 self.last_finalized_timestamp = block.timestamp();
+                // Announce pruned hashes
+                let pruned_hashes = self.prune(self.last_finalized_timestamp);
+                let pruned_count = pruned_hashes.total_pruned;
```
Comment (afck) [inline node/src/components/block_proposer.rs:272]: Do we know how fast pruning is? I wonder whether we should have a minimum delay, e.g. something like: if we already pruned within the last minute, don't prune again? But maybe it's not needed.

File: node/src/components/block_proposer.rs (hunk 2 above)
Comment (darthsiroftardis) [reply node/src/components/block_proposer.rs:253]: @afck Do you have an idea of what the log message should be like,

"block request timestamp is behind the timestamp of the last finalized block"?

Comment (afck) [reply node/src/components/block_proposer.rs:253]: Yes, sounds good. Maybe add something like: "not proposing a block; relevant deploys might have been purged"

**Review by Fraser999 — CHANGES_REQUESTED** [review]
As discussed, this PR should include tests for this change.

**Comment by goral09** [comment]
@Fraser999 @darthsiroftardis , Do we plan on including this PR with 1.5.0 release?

**Pull request closed without merging**